### PR TITLE
Navigation & Tenscript

### PR DESCRIPTION
**Pull request by geralddejong**
Headless revived
Tenscript improvements
Tinker mode
Navigation left and right

### CODE REVIEW

File: src/build/tenscript/build_phase.rs [hunk 1]
```diff
@@ -217,49 +215,49 @@ impl BuildPhase {
...
-                self.execute_node(fabric, IdentifiedFace { face_id }, &node, vec![]);
+                Self::execute_node(fabric, IdentifiedFace { face_id }, &node, vec![]);
             buds.extend(node_buds);
             marks.extend(node_marks);
         };
         (buds, marks)
     }
 
-    fn execute_node(&self, fabric: &mut Fabric, launch: Launch, node: &BuildNode, faces: Vec<UniqueId>) -> (Vec<Bud>, Vec<FaceMark>) {
+    fn execute_node(fabric: &mut Fabric, launch: Launch, node: &BuildNode, faces: Vec<UniqueId>) -> (Vec<Bud>, Vec<FaceMark>) {
```
Comment (mitchelljustin) [inline src/build/tenscript/build_phase.rs:225]: Cool

File: src/build/tenscript/build_phase.rs [hunk 2]
```diff
@@ -217,49 +215,49 @@ impl BuildPhase {
...
                     let (new_buds, new_marks) =
-                        self.execute_node(fabric, NamedFace { face_alias: branch_face_alias }, branch_node, brick_faces.clone());
+                        Self::execute_node(fabric, NamedFace { face_alias: branch_face_alias }, branch_node, brick_faces.clone());
                     buds.extend(new_buds);
                     marks.extend(new_marks);
                 }
             }
             Mark { mark_name } => {
-                let face_id = Self::find_launch_face(launch, &faces, fabric).expect("cannot mark from scratch");
+                let face_id = Self::find_launch_face(&launch, &faces, fabric).expect("cannot mark from scratch");
```
Comment (mitchelljustin) [inline src/build/tenscript/build_phase.rs:249]: Do these expect()s function as unreachable!()s? Or would errors be better surfaced via Result<>s?

File: src/build/tenscript/library.scm [hunk 3]
```diff
@@ -71,34 +57,36 @@
...
+                (face (alias TopZ) (mark :chest-1))
+                (face (alias BotX) (mark :chest-2))
+                (face (alias BotY) (grow "....X..." (scale .93) (mark :hands)))
+                )))
+          (face (alias Top:Right)
+            (grow 2 (scale .9)
+              (branch (alias Omni)
+                (face (alias TopY) (mark :chest-1))
+                (face (alias BotZ) (mark :chest-2))
+                (face (alias BotX) (grow "....X..." (scale .93) (mark :hands)))
```
Comment (mitchelljustin) [inline src/build/tenscript/library.scm:78]: Nice alignment on the (alias)es

File: src/build/tenscript/plan_runner.rs [hunk 4]
```diff
@@ -43,12 +41,12 @@ impl PlanRunner {
         let (next_stage, countdown) = match self.stage {
             Initialize => {
                 self.build_phase.init(fabric);
-                (GrowApproach, 200)
+                (GrowApproach, 500)
```
Comment (mitchelljustin) [inline src/build/tenscript/plan_runner.rs:44]: A bit too fast eh :)

File: src/camera.rs [hunk 5]
```diff
@@ -133,42 +134,42 @@ pub enum Target {
...
-        }
+    pub fn ideal_camera_distance(&self, fabric: &Fabric) -> Option<f32> {
+        self.selected_face(fabric).map(|(_, face)| face.scale * 7.0)
     }
 
     pub fn allow_vertical_rotation(&self) -> bool {
+        self.selected_face_id().is_some()
+    }
+
+    pub fn selected_face<'a>(&self, fabric: &'a Fabric) -> Option<(UniqueId, &'a Face)> {
```
Comment (mitchelljustin) [inline src/camera.rs:164]: 😮 lifetimes

File: src/fabric/mod.rs [hunk 6]
```diff
@@ -116,7 +116,11 @@ impl Fabric {
     }
 
     pub fn face(&self, id: UniqueId) -> &Face {
-        self.faces.get(&id).expect(&format!("face not found {id:?}"))
+        self.faces.get(&id).unwrap_or_else(|| panic!("face not found {id:?}"))
+    }
+
+    pub fn newest_face_id(&self) -> UniqueId {
+        *self.faces.keys().into_iter().max().unwrap()
```
Comment (mitchelljustin) [inline src/fabric/mod.rs:123]: is faces guaranteed to not be empty?

File: src/fabric/mod.rs [hunk 7]
```diff
@@ -246,16 +250,14 @@ impl Fabric {
...
         id
     }
 }
 
-#[derive(Clone, Debug, Copy, PartialEq, Default, Hash, Eq)]
-pub struct UniqueId {
-    pub id: usize,
-}
+#[derive(Clone, Debug, Copy, PartialEq, Default, Hash, Eq, Ord, PartialOrd)]
+pub struct UniqueId(usize);
```
Comment (mitchelljustin) [inline src/fabric/mod.rs:260]: Nice!!

File: src/user_interface/keyboard.rs [hunk 8]
```diff
@@ -78,24 +76,26 @@ impl Component for Keyboard {
...
     pub fn current(&self) -> Menu {
         self.current.last().unwrap().clone()
     }
 
     pub fn key_pressed(&self, keycode_pressed: &VirtualKeyCode) -> (Vec<Menu>, Option<Action>) {
         let mut current = self.current.clone();
-        if keycode_pressed == &Escape {
-            current.clear();
-            current.push(self.menu.clone());
+        if keycode_pressed == &Escape || keycode_pressed == &Back {
```
Comment (mitchelljustin) [inline src/user_interface/keyboard.rs:95]: This reads nicer if you pass an owned VirtualKeyCode instead and compare without reference.

File: src/user_interface/keyboard.rs [hunk 9]
```diff
@@ -110,8 +110,9 @@ impl Keyboard {
                     return None;
                 }
                 if action.is_some() {
-                    current.clear();
-                    current.push(self.menu.clone());
+                    if current.len() > 1 {
+                        current.pop();
+                    }
```
Comment (mitchelljustin) [inline src/user_interface/keyboard.rs:115]:
```suggestion
                    current.pop();
```
Will just return `None` if current is empty, and won't panic

**Review by mitchelljustin — APPROVED** [review]

File: src/build/tenscript/build_phase.rs (hunk 1 above)
Comment (geralddejong) [reply src/build/tenscript/build_phase.rs:225]: This was whispered in my ear by the tooling.  Brilliant!

**Pull request merged**